Added ref card tests for the new security commands

Hunterness opened the pull request:
Added in manual modeling so they can become visible :) https://github.com/neo-technology/neo4j-manual-modeling/pull/994

Note: role management privileges links to section which is merged but not yet visible in the preview so it might currently be dead if merged

Hunterness committed:
Added ref card tests for the new security commands

Note: role management privileges links to section which is merged but
not yet visible in the preview so it might currently be dead

Hunterness commented on cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/UserManagementTest.scala:
```diff
@@ -0,0 +1,80 @@
+###
+
+List all users in the system, if they need to update their password, their status and roles.
+(★) Status and roles are enterprise only.
```
Maybe add a new line here to actually get this on a new line

Lojjs replied:
I think this can be interpreted as if both passwords, status and roles are things that might need update. I would suggest `List all users in the system, with their status, roles and if they need to change their password.
(★) Status and roles are enterprise only.` 

Also I think to add a newline in there is a good idea.

Lojjs commented on cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/DatabaseManagementTest.scala:
```diff
@@ -0,0 +1,85 @@
+CREATE OR REPLACE DATABASE myDatabase
+###
+
+(★) Create a database named `myDatabase`, if there already existed a database with that name then the old is deleted and a new one created.
```
I think this sentence is too long and reads a bit weird. I would prefer something like:
``(★) Create a database named `myDatabase`. If  a database with that name already existed, that one is deleted and a new one is created.``

Lojjs commented on cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/UserManagementTest.scala:
```diff
@@ -0,0 +1,80 @@
+
+ALTER USER alice SET PASSWORD $password CHANGE NOT REQUIRED
+###
+
```
I think it would be nice with some `SET STATUS SUSPENDED` example here as well (maybe as part of the ALTER USER example as it is enterprise only).

Lojjs commented on cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/RoleManagementTest.scala:
```diff
@@ -0,0 +1,83 @@
+SHOW POPULATED ROLES WITH USERS
+###
+
+List all roles, and their users, that are assigned to users in the system.
```
I find this text slightly confusing. Maybe `List all roles that are assigned to at least one user in the system and the users for each role.` is clearer? Or split it up in 2 rows, one with `SHOW POPULATED ROLES` and one for `SHOW ROLES WITH USERS`.

Hunterness replied:
I feel like this is a combo you might like to run so I don't think I'd like to split it, but can see if I can clarify.

Lojjs commented on cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/DatabaseManagementTest.scala:
```diff
@@ -0,0 +1,85 @@
+SHOW DATABASE myDatabase
+###
+
+List all the information about the database `myDatabase`.
```
I think `all the information` is not a good wording. Comparing with `SHOW DATABASES` it sounds like `SHOW DATABASE myDatabase` shows more info. Also probably the system has more information than it is showing, so it is not all available information. Why not just `List information about ...`

Lojjs commented on cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/DatabaseManagementTest.scala:
```diff
@@ -0,0 +1,85 @@
+SHOW DEFAULT DATABASE
+###
+
+List all the information about the current default database.
```
Same as above regarding `all the information`

Lojjs commented on cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/ListPrivilegeTest.scala:
```diff
@@ -0,0 +1,62 @@
+SHOW PRIVILEGES
+###
+
+List all privileges in the system, what privilege it is, what resource and segment its on and which role it is assigned to.
```
Since the resource and segment is also part of the privilege, it is unclear to me what `what privilege it is` refers to? Is it `GRANT/DENY` or `TRAVERSE/READ/...`. Since we have talked about changing the output of the `SHOW` commands and you are anyway just listing 4 of 6 columns, I would prefer a simpler description like `List all privileges in the system and which roles they are assigned to.`

Lojjs commented on cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/ListPrivilegeTest.scala:
```diff
@@ -0,0 +1,62 @@
+SHOW ROLE my_role PRIVILEGES
+###
+
+List all privileges belonging to the role `my_role`, what privilege it is and what resource and segment its on.
```
Of same reason as above, I would prefer just ``List all privileges belonging to the role `my_role` ``. Maybe also change the word `belonging` to `assigned`.

Lojjs commented on cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/ListPrivilegeTest.scala:
```diff
@@ -0,0 +1,62 @@
+SHOW USER alice PRIVILEGES
+###
+
+List all privileges belonging to the user `alice`, what privilege it is, what resource and segment its on and which role it is assigned to.
```
Of same reason as above, I would prefer just ``List all privileges belonging to the user `alice` and which role each privilege is assigned to.``

Lojjs commented on cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/GraphPrivilegeTest.scala:
```diff
@@ -0,0 +1,71 @@
+ */
+package org.neo4j.cypher.docgen.refcard
+
+class GraphPrivilegeTest extends AdministrationCommandTestBase {
```
I think you have done a pretty good job here to pick some representative examples. Only thing I would like to see is RELATIONSHIP or ELEMENT instead of NODES in some of the examples.

Lojjs commented on cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/GraphPrivilegeTest.scala:
```diff
@@ -0,0 +1,71 @@
+
+Grant `read` and `traverse` privileges to the role `my_role`.
+Here the `read` privilege is on all properties on all nodes with label `Label` on the graph `foo` and
+the `traverse` privilege on all nodes with label `Label` on the graph `foo`.
```
I think this text doesn't have to be so long. Something like  ``Grant `read` privilege on all properties and `traverse` privilege to the role `my_role`. Here both privileges apply to all nodes with label `Label` in the graph `foo`.`` would do.

Lojjs commented on cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/GraphPrivilegeTest.scala:
```diff
@@ -0,0 +1,71 @@
+DENY READ {prop} ON GRAPH foo NODES Label TO my_role
+###
+
+Deny `read` privilege on property `prop` on all nodes with label `Label` on the graph `foo` to the role `my_role`.
```
In documentation, we use `in the graph` rather than `on the graph`.

Lojjs commented on cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/DatabasePrivilegeTest.scala:
```diff
@@ -0,0 +1,153 @@
+
+Grant all database privileges on the database `baz` to the role `my_role`.
+This means that users with `my_role` can access, start and stop the database as well as
+create and drop indexes and constraints and create labels, relationship types and property keys on the database.
```
I think it is a bit risky to list all privileges here, since we might forget to update the list if we add any new database privileges in the future. Maybe it is better to just write ``Grant all database privileges above on the database `baz` to the role `my_role` ``? Or just mention a few examples in the text. What do you think?

Lojjs commented on cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/RoleManagementPrivilegeTest.scala:
```diff
@@ -0,0 +1,84 @@
+GRANT ROLE MANAGEMENT ON DBMS TO my_role
+###
+
+Grant the privileges to create, delete, assign, remove and show roles to the role `my_role`.
```
As with `ALL ON DATABASE`, it might be a bit risky to list all of the privileges explicitly in case we add new ones and forget to update this lst. In this case we do not have that many privileges though and I think it is less risk that we will add more, so maybe ok to keep it here.

Lojjs reviewed (changes requested):
Overall good work but I have a number of text change suggestions.

Hunterness committed:
Fixed review comments

Hunterness committed:
Updated security chapter on docs to not tell people wrong things

It was saying you could grant graph privileges to multiple graphs
at the same time using a comma separated list which is not supported

davidoliverSP2 commented on cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/GraphPrivilegeTest.scala:
```diff
@@ -0,0 +1,70 @@
+###assertion=update-one
+//
+
+DENY READ {prop} ON GRAPH foo RELATIONSHIP Type TO my_role
```
This doesn't appear to match the description - I wasn't sure whether to change the example or the description! Could you please amend?

Hunterness replied:
We had nodes first and then we have updated the query but not the description.

Since I have done some suggestions on other places I'll do that here as well.

davidoliverSP2 commented on cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/GraphPrivilegeTest.scala:
```diff
@@ -0,0 +1,70 @@
+###assertion=update-four
+//
+
+GRANT MATCH {*} ON GRAPH foo ELEMENTS Label TO my_role
```
The description doesn't quite match what we have in [Section 10.3.4.2 of the Operations Manual, about privileges](https://neo4j.com/docs/operations-manual/4.0/authentication-authorization/access-control/#_privileges_of_literal_researcher_literal), where it says  `GRANT MATCH {*}` will grant access to everything.

Hunterness replied:
The `MATCH {*}` will give `TRAVERSE` and `READ {*}` which I think are what the below text states (at least its supposed to). It is also what the text in the operations manual is trying to say but using the word `access` there might be slightly odd since we have a separate `ACCESS` privilege which is NOT included in `MATCH`.

The `MATCH {*}` will still be restricted by the label/type and graph from the command, here `GRAPH foo` and `ELEMENTS Label` (both which are missing from the operations manual example).

davidoliverSP2 committed:
edit the tests for consistency

davidoliverSP2 reviewed (changes requested):
Thanks @Hunterness 

@renetapopova and I have pushed a light edit - there were a couple of descriptions that didn't make sense to us - could you take a look at the questions below? Many thanks!

Hunterness commented on cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/DatabasePrivilegeTest.scala:
```diff
@@ -40,111 +40,111 @@ class DatabasePrivilegeTest extends AdministrationCommandTestBase {
 ###
 
-Revoke grant privilege to create new property keys on the database `bar` from the role `my_role`.
+Revoke the grant privilege from a role to create new property names on a specified database.
```
```suggestion
Revoke the grant privilege to create new property names on a specified database from a role.
```

Hunterness commented on cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/DatabasePrivilegeTest.scala:
```diff
@@ -40,111 +40,111 @@ class DatabasePrivilegeTest extends AdministrationCommandTestBase {
 ###
 
-Grant all database privileges above on the database `baz` to the role `my_role`.
+Grant all privileges on a specified database to a role.
```
Since this is only granting the Database privileges and not graph or dbms privileges maybe the `database` should stay?

```suggestion
Grant all database privileges on a specified database to a role.
```

Hunterness commented on cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/GraphPrivilegeTest.scala:
```diff
@@ -38,32 +38,32 @@ class GraphPrivilegeTest extends AdministrationCommandTestBase {
 ###
 
-Deny `read` privilege on property `prop` on all nodes with label `Label` in the graph `foo` to the role `my_role`.
+Deny `read` privilege on a specified property, on all nodes with label `Label` in the graph `foo`, to the role `my_role`.
```
```suggestion
Deny `read` privilege on a specified property, on all relationships with a specified type in a specified graph to a role.
```

Hunterness commented on cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/UserManagementTest.scala:
```diff
@@ -41,48 +41,48 @@ class UserManagementTest extends AdministrationCommandTestBase {
 ###
 
-(★) Change the status for the user `alice` so that she is suspended. Use `SET STATUS ACTIVE` to reactivate `alice`.
+(★) Change the user status to suspended. Use `SET STATUS ACTIVE` to reactivate the user name.
```
```suggestion
(★) Change the user status to suspended. Use `SET STATUS ACTIVE` to reactivate the user.
```

Hunterness commented on cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/UserManagementTest.scala:
```diff
@@ -41,48 +41,48 @@ class UserManagementTest extends AdministrationCommandTestBase {
 ###
 
-Change the password for the logged in user and update so that a password change is not required on next login.
+Change the password of the logged-in user, and update. The user will not be required to change this password on the next login.
```
```suggestion
Change the password of the logged-in user. The user will not be required to change this password on the next login.
```

Hunterness commented on cypher/refcard-tests/src/test/scala/org/neo4j/cypher/docgen/refcard/UserManagementTest.scala:
```diff
@@ -41,48 +41,48 @@ class UserManagementTest extends AdministrationCommandTestBase {
 
-List all users in the system, their status, roles and if they need to change their password.\n
-(★) Status and roles are enterprise only.
+List all users in the system, their status, roles and if they need to change their password. +
```
We wanted to have a line break here but did not know how to, don't know what the `+` does either :P

davidoliverSP2 replied:
It's the `+` that starts the new line 😄 

I'll see if I can add them where it makes sense to start new lines.

Hunterness reviewed (commented):
I looked at your changes and hade some comments/suggestions. I therefore addressed one of your comments with suggestion instead of amending in a fix. The other I answered but don't yet know what we should do with so lets see what you think.

davidoliverSP2 committed:
update the User Management section

davidoliverSP2 committed:
update the Graph privileges section

davidoliverSP2 committed:
update the Database privileges section

davidoliverSP2 reviewed (approved):
thanks @Hunterness 

I've amended the content as per your suggestions. I'm happy for this to be merged now. I noticed there are still a few comments from @Lojjs - please let me know if you have any further actions.

Hunterness commented:
@davidoliverSP2 I think that all of Louise comments were addressed before and that they just wasn't marked as resolved (since she approved it afterwards we see that as a marker that she has no further comments).

It will be nice to finally see this merged :) I guess you will do that?

davidoliverSP2 commented:
@Hunterness yes, I'm happy to merge 😄
